### PR TITLE
changefeedccl: add scoping support for max_behind_nanos

### LINKED ISSUE #132281: cdc: add checkpoint progress metrics per changefeed
The `changefeed.max_behind_nanos` metric currently only reflects the lag of the furthest behind changefeed. However, there are times when it's useful to differentiate changefeed's progress. For example, if a cluster has multiple lagging changefeeds or a customer wants to check if a specific changefeed is lagging via a scoped metric label. We should make a scoped version of this metric.

Jira issue: CRDB-42917

Epic CRDB-42868

### PR DESCRIPTION
**Pull request by aerfrei**
Currently, changefeed.max_behind_nanos is a measurement of the
lag of the furthest behind changefeed. We'd like to be able to
support scoping/grouping changefeeds. This applies scoping to
that metric similar to the scoping on changefeed.aggregator_progress.

Epic: none
Fixes: https://github.com/cockroachdb/cockroach/issues/132281

Release note (ops change): the changefeed.max_behind_nanos metric
now supports scoping with metric labels.

### CODE REVIEW

File: pkg/ccl/changefeedccl/changefeed_processors.go [hunk 1]
```diff
@@ -1467,21 +1459,8 @@ func (cf *changeFrontier) close() {
...
 
-// closeMetrics de-registers from the progress registry that powers
-// `changefeed.max_behind_nanos`. This method is idempotent.
 func (cf *changeFrontier) closeMetrics() {
-	// Delete this feed from the MaxBehindNanos metric so it's no longer
-	// considered by the gauge.
-	func() {
-		cf.metrics.mu.Lock()
-		defer cf.metrics.mu.Unlock()
-		if cf.metricsID > 0 {
```
Comment (aerfrei) [inline pkg/ccl/changefeedccl/changefeed_processors.go:1478]: It seems to me that without this check this function will no longer be idempotent which seems dangerous. Open to suggestions about what to do here. Maybe this is reason to keep cf.metrics.mu around?

**Review by andyyang890 — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r1, 2 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137534#-:-OEPwZOx4JHKHuLe9uIX:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @aerfrei, @arjunmahishi, @asg0451, @rharding6373, and @xinhaoz)

___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137534#-OEPtqRO-knLPCiUh4qj:-OEPtqRPA12R17rcZxs6:b-yqhm3g):*
FYI it's not necessary to put an Epic if you've already linked an issue (linking an issue is preferred per https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/2009039063/Issue+and+Epic+Refs+in+Commits+and+PRs)
___
*[`pkg/ccl/changefeedccl/metrics.go` line 1003 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137534#-OEPv7gc6E89G70lAKj3:-OEPv7gc6E89G70lAKj4:bipdn5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e1a766de3415a9f7acc8e18174e62193d81ac6d/pkg/ccl/changefeedccl/metrics.go#L1003)):*
> ```Go
> 		var max int64
> 		for _, val := range childValues {
> 			if val != 0 && val > max {
> ```

Do you need the `val != 0` clause? Another thing you could do is leverage go's new max function: `return max(0, childValues...)`

**Review by andyyang890 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137534)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @aerfrei, @arjunmahishi, @asg0451, @rharding6373, and @xinhaoz)

___
*[`pkg/ccl/changefeedccl/metrics.go` line 1208 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137534#-OEPxBZt17YEePVtf1ai:-OEPxBZt17YEePVtf1aj:bernhs2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e1a766de3415a9f7acc8e18174e62193d81ac6d/pkg/ccl/changefeedccl/metrics.go#L1208)):*
> ```Go
> 				return 0
> 			}
> 			return timeutil.Now().UnixNano() - minTs
> ```

Is there a reason you chose to do a raw subtraction instead of using the `Sub` method? From skimming the code, it seems like we usually convert the `hlc.Timestamp` to a go time and then use `Sub`.

**Review by aerfrei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137534)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @andyyang890, @arjunmahishi, @asg0451, @rharding6373, and @xinhaoz)

___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137534#-OEPtqRO-knLPCiUh4qj:-OEQ6HgG-rKo_6fMuglZ:b-rj98yq):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

FYI it's not necessary to put an Epic if you've already linked an issue (linking an issue is preferred per https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/2009039063/Issue+and+Epic+Refs+in+Commits+and+PRs)
</blockquote></details>

I'll change this, thanks for the tip!
___
*[`pkg/ccl/changefeedccl/metrics.go` line 1003 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137534#-OEPv7gc6E89G70lAKj3:-OEQ61bV2B01wWJ0sDV0:b-e5vybh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e1a766de3415a9f7acc8e18174e62193d81ac6d/pkg/ccl/changefeedccl/metrics.go#L1003)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Do you need the `val != 0` clause? Another thing you could do is leverage go's new max function: `return max(0, childValues...)`
</blockquote></details>

Good point on the val check, I'll remove that. I get a compilation error on using max that way, seems like max doesn't support it. https://stackoverflow.com/questions/78778577/a-compilation-error-while-using-builtin-func-max-with-go-1-22-1-invalid-ope
___
*[`pkg/ccl/changefeedccl/metrics.go` line 1208 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137534#-OEPxBZt17YEePVtf1ai:-OEQ5AXT3dTJmneAPI-2:b-oemoe0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e1a766de3415a9f7acc8e18174e62193d81ac6d/pkg/ccl/changefeedccl/metrics.go#L1208)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Is there a reason you chose to do a raw subtraction instead of using the `Sub` method? From skimming the code, it seems like we usually convert the `hlc.Timestamp` to a go time and then use `Sub`.
</blockquote></details>

The reason I did it this way is because I chose to reuse the minTimestampGetter, which returns an int64. If I still wanted to use it, I considered converting back into a timestamp to do the subtraction and then back into nanoseconds like this: "timeutil.Now().Sub(time.Unix(0, minTs)).Nanoseconds()" which seems less readable and I didn't see a strong argument the other way. Thoughts?

**Review by andyyang890 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137534)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @aerfrei, @arjunmahishi, @asg0451, @rharding6373, and @xinhaoz)

___
*[`pkg/ccl/changefeedccl/metrics.go` line 1003 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137534#-OEPv7gc6E89G70lAKj3:-OEQGuws1MvkL2SihRkR:b-l376m9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e1a766de3415a9f7acc8e18174e62193d81ac6d/pkg/ccl/changefeedccl/metrics.go#L1003)):*
<details><summary><i>Previously, aerfrei (Aerin Freilich) wrote…</i></summary><blockquote>

Good point on the val check, I'll remove that. I get a compilation error on using max that way, seems like max doesn't support it. https://stackoverflow.com/questions/78778577/a-compilation-error-while-using-builtin-func-max-with-go-1-22-1-invalid-ope
</blockquote></details>

Ah ok. In that case, maybe you could use `slices.Max` like the StackOverflow answer says instead?
___
*[`pkg/ccl/changefeedccl/metrics.go` line 1208 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137534#-OEPxBZt17YEePVtf1ai:-OEQJ2GiAP7YRCrYh7mI:b-w344dp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e1a766de3415a9f7acc8e18174e62193d81ac6d/pkg/ccl/changefeedccl/metrics.go#L1208)):*
<details><summary><i>Previously, aerfrei (Aerin Freilich) wrote…</i></summary><blockquote>

The reason I did it this way is because I chose to reuse the minTimestampGetter, which returns an int64. If I still wanted to use it, I considered converting back into a timestamp to do the subtraction and then back into nanoseconds like this: "timeutil.Now().Sub(time.Unix(0, minTs)).Nanoseconds()" which seems less readable and I didn't see a strong argument the other way. Thoughts?
</blockquote></details>

Ah ok, that makes sense to me.

**Review by aerfrei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137534)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @andyyang890, @arjunmahishi, @asg0451, @rharding6373, and @xinhaoz)

___
*[`pkg/ccl/changefeedccl/metrics.go` line 1003 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137534#-OEPv7gc6E89G70lAKj3:-OEQZEtQ0H-RCKkGxtEr:b-e55spp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e1a766de3415a9f7acc8e18174e62193d81ac6d/pkg/ccl/changefeedccl/metrics.go#L1003)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Ah ok. In that case, maybe you could use `slices.Max` like the StackOverflow answer says instead?
</blockquote></details>

Great, done!

**Review by andyyang890 — APPROVED** [review]

**Review by asg0451 — APPROVED** [review]

File: pkg/ccl/changefeedccl/metrics.go [hunk 2]
```diff
@@ -986,6 +977,16 @@ func newAggregateMetrics(histogramWindow time.Duration, lookup *cidr.Lookup) *Ag
 		Measurement: "Count",
 		Unit:        metric.Unit_COUNT,
 	}
+	// TODO(dan): This was intended to be a measure of the minimum distance of
+	// any changefeed ahead of its gc ttl threshold, but keeping that correct in
+	// the face of changing zone configs is much harder, so this will have to do
+	// for now.
+	metaChangefeedMaxBehindNanos := metric.Metadata{
+		Name:        "changefeed.max_behind_nanos",
+		Help:        "(Deprecated in favor of checkpoint_progress) The most any changefeed's persisted checkpoint is behind the present",
```
[the remark this answers is not in the document]
Comment (aerfrei) [reply pkg/ccl/changefeedccl/metrics.go]: Done.

**Review by aerfrei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137534)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @andyyang890, @arjunmahishi, @asg0451, @rharding6373, and @xinhaoz)

**Comment by aerfrei** [comment]
bors r=andyyang890,asg0451

**Comment by aerfrei** [comment]
blathers backport 23.2 24.1 24.2 24.3

**Comment by aerfrei** [comment]
blathers backport 23.2 24.1 24.2 24.3

**Pull request merged**